### PR TITLE
Cache "global" trait resolutions

### PR DESCRIPTION
**Pull request by nikomatsakis**
When we successfully resolve a trait reference with no type/lifetime parameters, like `i32: Foo` or `Box<u32>: Sized`, this is in fact globally true. This patch adds a simple global to the tcx to cache such cases. The main advantage of this is really about caching things like `Box<Vec<Foo>>: Sized`. It also points to the need to revamp our caching infrastructure -- the current caches make selection cost cheaper, but we still wind up paying a high cost in the confirmation process, and in particular unrolling out dependent obligations. Moreover, we should probably do caching more uniformly and with a key that takes the where-clauses into account. But that's for later.

For me, this shows up as a reasonably nice win (20%) on Servo's script crate (when built in dev mode). This is not as big as my initial measurements suggested, I think because I was building my rustc with more debugging enabled at the time. I've not yet done follow-up profiling and so forth to see where the new hot spots are. Bootstrap times seem to be largely unaffected.

cc @pcwalton 

This is technically a [breaking-change] in that functions with unsatisfiable where-clauses may now yield errors where before they may have been accepted. Even before, these functions could never have been _called_ by actual code. In the future, such functions will probably become illegal altogether, but in this commit they are still accepted, so long as they do not rely on the unsatisfiable where-clauses. As before, the functions still cannot be called in any case.

### CODE REVIEW

**Comment by metajack** [comment]
cc @metajack

File: src/test/compile-fail/cross-fn-cache-hole.rs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+trait Bar<X> { }
+
+fn vacuous<A>()
+    where i32: Foo<u32, A>
+{
+    // vacuous could never be called, because it requires that i32:
+    // Bar<u32>. But the code doesn't check that this could never be
+    // satisfied.
+    require::<i32, u32>();
+    //~^ ERROR the trait `Bar<u32>` is not implemented for the type `i32`
```
Comment (eddyb) [inline src/test/compile-fail/cross-fn-cache-hole.rs]: Is this a "breaking change" by any means?

File: src/test/compile-fail/unsized5.rs [hunk 2]
```diff
@@ -24,7 +24,7 @@ struct S3 {
     g: [usize]
 }
 struct S4 {
-    f: str, //~ ERROR `core::marker::Sized` is not implemented
+    f: [u8], //~ ERROR `core::marker::Sized` is not implemented
```
Comment (arielb1) [inline src/test/compile-fail/unsized5.rs]: I think I would try to do this in a way that doesn't worsen error messages much.

File: src/test/compile-fail/cross-fn-cache-hole.rs (hunk 1 above)
Comment (arielb1) [inline src/test/compile-fail/cross-fn-cache-hole.rs]: Kind of. This function *does* compile uselessly in stable/nightly.

File: src/librustc/middle/ty.rs [hunk 3]
```diff
@@ -753,6 +753,11 @@ pub struct ctxt<'tcx> {
     /// for things that do not have to do with the parameters in scope.
     pub selection_cache: traits::SelectionCache<'tcx>,
 
+    /// A set of predicates that have been fulfilled *somewhere*.
+    /// This is used to avoid duplicate work. Predicates are only
+    /// added to this set when they
```
Comment (nrc) [inline src/librustc/middle/ty.rs]: Missing the end of this comment?

File: src/librustc/middle/ty.rs [hunk 4]
```diff
@@ -815,6 +820,11 @@ bitflags! {
         const HAS_TY_ERR         = 1 << 6,
         const HAS_PROJECTION     = 1 << 7,
         const HAS_TY_CLOSURE     = 1 << 8,
+
+        // true if there are "names" of types and regions and so forth
+        // that are local to a particular fn
+        const HAS_LOCAL_NAMES   = 1 << 8,
```
Comment (nrc) [inline src/librustc/middle/ty.rs]: should be `1 << 9`

**Comment by arielb1** [comment]
Translation creates tons of these useless nested obligations - I think the primary benefits may be there. Could you post some performance numbers?

File: src/librustc/middle/ty.rs (hunk 4 above)
Comment (eddyb) [inline src/librustc/middle/ty.rs]: From other parts of the patch, it looked like a rename from `HAS_TY_CLOSURE` to `HAS_LOCAL_NAMES`.

**Comment by arielb1** [comment]
Anyway, rustc perf numbers:

```
Before:
548.54user 4.54system 7:16.79elapsed 126%CPU (0avgtext+0avgdata 1174632maxresident)k

After:
537.65user 4.22system 7:04.81elapsed 127%CPU (0avgtext+0avgdata 1175460maxresident)k
```

This is a 2%-3% improvement on rustc. It is indeed mostly a trans improvement (improves translation from 89.6s to 79.3s, but type-checking from 62.6s to 59s) - probably because of workload differences between rustc and trans. Still nice.

File: src/librustc/middle/ty.rs [hunk 5]
```diff
@@ -3018,11 +3066,12 @@ impl FlagComputation {
             }
 
             &TyClosure(_, substs) => {
-                self.add_flags(TypeFlags::HAS_TY_CLOSURE);
+                self.add_flags(TypeFlags::HAS_LOCAL_NAMES);
```
Comment (nrc) [inline src/librustc/middle/ty.rs]: This looks to be a mistake - at least the removal.

**Comment by nrc** [comment]
r=me with the rebasing errors fixed

**Comment by nikomatsakis** [comment]
@arielb1 time numbers for compiling script (in dev mode) on my machine:

**After:**

- real    2m2.024s
- user    1m58.926s
- sys     0m3.098s

**Before:**

- real    2m35.145s
- user    2m32.113s
- sys     0m2.980s

Regarding error messages, I discussed it with @pcwalton, and we felt like it was worth landing this sooner and dealing with error messages later -- though I left a note for what I think is the best fix. Doesn't seem *too* hard, I could do some local experiments.

**Comment by nikomatsakis** [comment]
@arielb1 based on the profiling I did, the benefits were not just in trans, but they were primarily due to Sized and Copy bounds. However, the numbers I see now look quite different than what I was seeing then, so I ought to re-run the profiles (I think this is due to how I compiled rustc, but I'm not sure).

**Comment by arielb1** [comment]
@nikomatsakis 

I feel like a better way of doing this is to do each global predicate in a fresh "global" fulfillcx, and merge duplicates if fulfillment is successful (I don't think it is worth to keep track of errors). I don't like the idea of polluting a global cache with potentially-bogus results - it seems to me like it would easily cause mysterious ICE-s.

**Comment by nikomatsakis** [comment]
@arielb1 well, it doesn't really matter so much what fulfill cx we do it in, the problem is more that we have to track when all dependent resolutions are complete. I do agree there is the potential for ICEs here.

**Comment by arielb1** [comment]
This does indeed seem to produce 13-14% improvements in both typeck and trans time when assertions are enabled (I guess these `debug!`-s add up).

**Pull request merged**